File: Day353/SplitArrayLargestSum.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            
            String St[] = read.readLine().split(" ");
            
            int N = Integer.parseInt(St[0]);
            int K = Integer.parseInt(St[1]);
            
            String S[] = read.readLine().split(" ");
            
            int[] arr = new int[N];
            
            for(int i=0 ; i<N ; i++)
                arr[i] = Integer.parseInt(S[i]);

            Solution ob = new Solution();
            System.out.println(ob.splitArray(arr,N,K));
        }
    }
}
// } Driver Code Ends


class Solution {
    public static int solve(int arr[], int n, long max) {
        int count  =1; 
        long curr = 0;
        for (int num: arr) {
            curr += num;
            if (curr > max) {
                count++;
                curr = num;
            }
        }
        return count;
    }
    static int splitArray(int[] arr , int N, int K) {
        // code here
        long l = Arrays.stream(arr).max().getAsInt();
        long h = Arrays.stream(arr).sum();
        while (l < h) {
            long mid = l + (h - l) /2 ;
            int temp = solve(arr, N, mid);
            if (temp > K) {
                l = mid + 1;
            } else {
                h = mid;
            }
        }
        return (int)l;
    }
};

File: Day353/SplitArrayLargestSum.go
package main

import (
	"fmt"
	"sort"
)

func solve(arr []int, n int, max int64) int {
	count := 1
	curr := int64(0)

	for _, num := range arr {
		curr += int64(num)
		if curr > max {
			count++
			curr = int64(num)
		}
	}

	return count
}

func splitArray(arr []int, N, K int) int {
	var l, h int64
	for _, num := range arr {
		l = max(l, int64(num))
		h += int64(num)
	}

	for l < h {
		mid := l + (h-l)/2
		temp := solve(arr, N, mid)
		if temp > K {
			l = mid + 1
		} else {
			h = mid
		}
	}

	return int(l)
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	var t int
	fmt.Scan(&t)

	for t > 0 {
		var N, K int
		fmt.Scan(&N, &K)

		arr := make([]int, N)
		for i := 0; i < N; i++ {
			fmt.Scan(&arr[i])
		}

		res := splitArray(arr, N, K)
		fmt.Println(res)

		t--
	}
}
